### PR TITLE
Initial implementation of partitions (Create statement)

### PR DESCRIPTION
**Pull request by Thirumalai-Shaktivel**
Signed-off-by: Thirumalai-Shaktivel <thirumalaishaktivel@gmail.com>

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

### CODE REVIEW

**Comment by Thirumalai-Shaktivel** [comment]
Anything else I need to fix?? If not, then this is good to go in... (this PR is completed from my side, I will send another PR for partition_definition and subpartition_definition)

File: go/vt/sqlparser/sql.y [hunk 1]
```diff
@@ -769,10 +780,11 @@ set_session_or_global:
   }
 
 create_statement:
-  create_table_prefix table_spec
+  create_table_prefix table_spec partitions_options_opt
```
Comment (GuptaManan100) [inline go/vt/sqlparser/sql.y:783]: We should move this inside the table_spec rule

File: go/vt/sqlparser/ast.go [hunk 2]
```diff
@@ -464,13 +464,14 @@ type (
...
-		Comments    Comments
-		FullyParsed bool
+		Temp            bool
+		Table           TableName
+		IfNotExists     bool
+		TableSpec       *TableSpec
+		OptLike         *OptLike
+		Comments        Comments
+		FullyParsed     bool
+		PartitionOption *PartitionOption
```
Comment (GuptaManan100) [inline go/vt/sqlparser/ast.go:474]: Instead of creating a field here, could you move it inside the TableSpec field

File: go/vt/sqlparser/sql.y [hunk 3]
```diff
@@ -2271,6 +2283,136 @@ alter_statement:
     }
   }
 
+partitions_options_opt:
+  {
+    $$ = nil
+  }
+| PARTITION BY linear_opt id_or_var '(' expression ')' partitions_opt
```
Comment (GuptaManan100) [inline go/vt/sqlparser/sql.y:2290]: Why are we using id_or_var here, instead of the keyword `Hash`

File: go/vt/sqlparser/sql.y [hunk 4]
```diff
@@ -2271,6 +2283,136 @@ alter_statement:
...
+        SubPartition: $6,
+        Definitions: $7,
+      }
+    }
+
+subpartition_opt:
+  {
+    $$ = nil
+  }
+| SUBPARTITION BY linear_opt id_or_var '(' expression ')' subpartitions_opt
```
Comment (GuptaManan100) [inline go/vt/sqlparser/sql.y:2331]: Same comment. Is there a reason to use id_or_var instead of `Hash` keyword

**Review by GuptaManan100 — CHANGES_REQUESTED** [review]

**Comment by GuptaManan100** [comment]
Apart from these requested changes, everything looks good to me! Great work

File: go/vt/sqlparser/sql.y (hunk 3 above)
Comment (Thirumalai-Shaktivel) [reply go/vt/sqlparser/sql.y:2290]: The test fails(15) if we use the keyword `HASH`, so I used `id_or_var`
```
=== RUN   TestValid/alter_vschema_on_a_add_vindex_hash_(id)
    parse_test.go:2066: Parsing failed. 
        Expected/Got:
        alter vschema on a add vindex hash (id)
        alter vschema on a add vindex `hash` (id)
    --- FAIL: TestValid/alter_vschema_on_a_add_vindex_hash_(id) (0.00s)
```

Comment (GuptaManan100) [reply go/vt/sqlparser/sql.y:2290]: You should change the expected output of the test and keep `Hash`

**Comment by Thirumalai-Shaktivel** [comment]
Hi @GuptaManan100, thanks for the review. I made all the changes as you suggested...

**Review by GuptaManan100 — APPROVED** [review]
This looks good to me! 😺

**Comment by GuptaManan100** [comment]
The unit tests seem to be failing, could you take a look

**Comment by Thirumalai-Shaktivel** [comment]
It seems all the tests are passing in my local system, I don't know where the issue is!!

**Comment by harshit-gangal** [comment]
> It seems all the tests are passing in my local system, I don't know where the issue is!!

can you rebase from main and push it

**Comment by Thirumalai-Shaktivel** [comment]
Looks good to me!

**Review by GuptaManan100 — APPROVED** [review]

**Review by shlomi-noach — APPROVED** [review]
LGTM all things parsing! Nice work!

**Comment by Thirumalai-Shaktivel** [comment]
Thanks for reviewing and merging the code!!

**Pull request merged**